src/v/cloud_storage/inventory/inv_consumer.h

```diff
@@ -0,0 +1,117 @@
+/// in the report is expected to contain a path to a segment in the bucket for
+/// which the inventory was generated. A set of NTPs is provided to the consumer
+/// for validation, and only paths belonging to these NTPs are processed. For
+/// each path, its 64 bit hash is calculated and stored in memory. Once the
+/// cumulative size in memory of the stored hashes grows over a prescribed
+/// limit, the hashes for the invidividual NTPs are written to disk until the
+/// cumulative memory usage drops below the desired limit. On completion of the
+/// report consumption, all hashes in memory are flushed to disk.
+class inventory_consumer {
+    using write_all_t = ss::bool_class<struct write_all_tag>;
+
+public:
+    inventory_consumer(const inventory_consumer&) = delete;
+    inventory_consumer& operator=(const inventory_consumer&) = delete;
+    inventory_consumer(inventory_consumer&&) = delete;
+    inventory_consumer& operator=(inventory_consumer&&) = delete;
+
+    inventory_consumer(
+      std::filesystem::path hash_store_path,
+      absl::node_hash_set<model::ntp> ntps,
+      size_t max_hash_size_in_memory);
+
+    /// Consumes the report provided as input stream and populates in memory
+    /// path hashes. A flush to disk may be performed if the size of hashes
+    /// exceeds the prescribed limit.
+    ss::future<>
+    consume(ss::input_stream<char> s, gzip_compressed_t is_report_compressed);
+
+    static std::optional<model::ntp> ntp_from_path(std::string_view path);
+    static std::vector<ss::sstring> parse_row(ss::sstring row);
+
+    /// Ensures that streams opened by consumer are closed. Called during
+    /// exception handling if an exception is thrown during a stream
+    /// consumption.
+    ss::future<> stop();
+
+    ~inventory_consumer();
+
+private:
+    ss::future<> do_consume(
+      ss::input_stream<char> s, gzip_compressed_t is_report_compressed);
+
+    // Checks if the path belongs to one of the NTPs whose leadership belongs to
+    // this node. If so, the path is hashed and added to the `_path_hashes` map.
+    ss::future<> process_path(ss::sstring path);
+
+    // Writes the largest hash vectors to disk. The vectors are written in files
+    // named after their NTP. If write_all_entries is true, all hashes are
+    // written to disk. If it is not true, we write the vectors until total
+    // memory used falls below the max hash size limit.
+    ss::future<> flush(write_all_t write_all_entries = write_all_t::no);
+
+    // Writes hashes for a single NTP to it's data file. The data file is
+    // located in path: namespace/topic/partition_id. The parent directories are
+    // created if missing.
+    ss::future<>
+    flush_ntp_hashes(model::ntp ntp, fragmented_vector<uint64_t> hashes);
+    ss::future<> write_hashes_to_file(
+      ss::output_stream<char>& stream, fragmented_vector<uint64_t> hashes);
+
+    // The root directory for hash files
+    std::filesystem::path _hash_store_path;
+
```

Comment:
I wonder if it makes sense to encapsulate the details of whatever is spilling to disk into some ntp path store, and have the inv consume be a user of it, so this code can focus only on parsing the file rather than also managing on-disk naming scheme, etc